File: leetcode/python/p2381.py
class Solution(object):
    """
    https://leetcode.com/problems/shifting-letters-ii/
    """

    def shiftingLetters(self, s, shifts):
        """
        :type s: str
        :type shifts: List[List[int]]
        :rtype: str
        """
        n = len(s)
        prefix_sum = [0] * (n + 1)
        for shift in shifts:
            start, end, direction = shift
            if direction == 0:
                prefix_sum[start] -= 1
                prefix_sum[end + 1] += 1
            else:
                prefix_sum[start] += 1
                prefix_sum[end + 1] -= 1
        for i in range(1, n + 1):
            prefix_sum[i] += prefix_sum[i - 1]
        result = []
        for i in range(n):
            result.append(
                chr((ord(s[i]) - ord('a') + prefix_sum[i]) % 26 + ord('a')))
        return ''.join(result)


if __name__ == '__main__':
    s = "abc"
    shifts = [[0, 1, 0], [1, 2, 1], [0, 2, 1]]
    print(Solution().shiftingLetters(s, shifts))

File: leetcode/p2381.cpp


class Solution {
   public:
    /*
     * https://leetcode.com/problems/shifting-letters-ii/
     */

    std::string shiftingLetters(std::string s, std::vector<std::vector<int>>& shifts) {
        int n = s.length();
        std::vector<int> prefix_sum(n + 1, 0);

        for (const std::vector<int>& shift : shifts) {
            int start = shift[0];
            int end = shift[1];
            int direction = shift[2];

            if (direction == 0) {
                prefix_sum[start] -= 1;
                prefix_sum[end + 1] += 1;
            } else {
                prefix_sum[start] += 1;
                prefix_sum[end + 1] -= 1;
            }
        }

        for (int i = 1; i <= n; i++) {
            prefix_sum[i] += prefix_sum[i - 1];
        }

        std::string result;
        for (int i = 0; i < n; i++) {
            int shift_amount = (s[i] - 'a' + prefix_sum[i]) % 26;
            if (shift_amount < 0) shift_amount += 26;
            result += (char)(shift_amount + 'a');
        }

        return result;
    }
};
